`SELFDESTRUCT` fix for Cancun

OlivierBBB opened the pull request:
- fixes `SELFDESTRUCT` for Cancun
- implements more lenient upper limits for `NUMBER` and `TIMESTAMP` in `BLOCK_DATA`

OlivierBBB committed:
fix (?)

OlivierBBB committed:
feat: unit tests must run LONDON by default

OlivierBBB committed:
Merge branch 'arith-dev' into SELFDESCTUCT-for-Cancun-fix

OlivierBBB committed:
ras

OlivierBBB committed:
fix: accomodate for new type annotations in TXN_DATA module

OlivierBBB committed:
fix: TIMESTAMP no longer negative + bigger upper bounds

OlivierBBB committed:
Merge branch 'arith-dev' into SELFDESCTUCT-for-Cancun-fix

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/blockdata/moduleOperation/BlockdataOperation.java:
```diff
@@ -59,7 +59,7 @@ public abstract class BlockdataOperation extends ModuleOperation {
   private final BlockHeader prevBlockHeader;
   private static final EWord POWER_256_20 = EWord.of(TWOFIFTYSIX_TO_THE_TWENTY);
-  private static final EWord POWER_256_6 = EWord.of(BigInteger.ONE.shiftLeft(6 * 8));
+  private static final EWord POWER_256_8 = EWord.of(BigInteger.ONE.shiftLeft(8 * 8));
```
could use the same const as in the lisp as we do with `private static final EWord POWER_256_20 = EWord.of(TWOFIFTYSIX_TO_THE_TWENTY);`

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/systemTransaction/EIP4788BeaconBlockRootSection.java:
```diff
@@ -74,7 +77,9 @@ public EIP4788BeaconBlockRootSection(
 
     if (isNonTrivialOperation) {
-      final EWord keyTimestamp = EWord.of(timestamp % HISTORY_BUFFER_LENGTH);
+      final BigInteger timestampMod8191 =
```
the modulo8191 is a short, not a BigInt

OlivierBBB replied:
I did this because there were tests with very large TIMESTAMP which got interpreted as a negative number. And even the remainder was interpreted that way ...

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/systemTransaction/EIP4788BeaconBlockRootSection.java:
```diff
@@ -50,7 +53,7 @@ public EIP4788BeaconBlockRootSection(
     final AccountSnapshot beaconrootAccount =
         AccountSnapshot.canonical(hub, world, EIP4788_BEACONROOT_ADDRESS, false);
-    final long timestamp = blockHeader.getTimestamp();
+    final BigInteger timestamp = longToUnsignedBigInteger(blockHeader.getTimestamp());
```
ok for now, but we'll have to change this when I'm done with the Besu PR: https://github.com/Consensys/linea-tracer/issues/2241

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/txndata/rows/hubRows/HubRow.java:
```diff
@@ -37,8 +37,9 @@ public void traceRow(Trace.Txndata trace) {
         .pHubBtcBlockNumber(header.getNumber())
-        .pHubBtcBlockGasLimit(header.getGasLimit())
-        .pHubBtcBasefee(header.getBaseFee().get().getAsBigInteger().longValueExact())
+        .pHubBtcBlockGasLimit(Bytes.ofUnsignedLong(header.getGasLimit()))
```
what happens if gasLimit is `0xFFFFFFFFFFFFFFFF` ?

OlivierBBB replied:
I imagine it works as intended, no ? since it gets it as an unsigned integer rather than -1 as java would.

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/txndata/rows/hubRows/HubRow.java:
```diff
@@ -37,8 +37,9 @@ public void traceRow(Trace.Txndata trace) {
-        .pHubBtcBasefee(header.getBaseFee().get().getAsBigInteger().longValueExact())
+        .pHubBtcBlockGasLimit(Bytes.ofUnsignedLong(header.getGasLimit()))
+        .pHubBtcBasefee(
+            Bytes.ofUnsignedLong(header.getBaseFee().get().getAsBigInteger().longValueExact()))
```
betterUse bigIntToBytes, less conversion

letypequividelespoubelles commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/txndata/rows/hubRows/HubRowForUserTransactions.java:
```diff
@@ -44,9 +44,9 @@ public void traceRow(Trace.Txndata trace) {
         .pHubNonce(Bytes.ofUnsignedLong(txn.getBesuTransaction().getNonce()))
         .pHubValue(bigIntegerToBytes(txn.getBesuTransaction().getValue().getAsBigInteger()))
-        .pHubGasLimit(txn.getBesuTransaction().getGasLimit())
+        .pHubGasLimit(Bytes.ofUnsignedLong(txn.getBesuTransaction().getGasLimit()))
```
why having a :i128 in the constraints if in any case it's stored as a long in besu ? Same remark for all other typing change.

letypequividelespoubelles reviewed (commented):
LGTM for the selfdestruct, some concerns about the type change, I don't see why we have :i128 if besu stores info as long

OlivierBBB committed:
Merge branch 'arith-dev' into SELFDESCTUCT-for-Cancun-fix

OlivierBBB commented:
Merging early and addressing comments in future PR, see
- https://github.com/Consensys/linea-tracer/issues/2261
- https://github.com/Consensys/linea-constraints/issues/751

OlivierBBB committed:
constraints commit

DavePearce committed:
remove broken static import

OlivierBBB committed:
fix: MXPX_THRESHOLD   ->   CANCUN_MXPX_THRESHOLD

OlivierBBB committed:
Merge branch 'arith-dev' into SELFDESCTUCT-for-Cancun-fix
